pinot-core/src/main/java/org/apache/pinot/core/common/DataBlockCache.java

```diff
@@ -120,6 +121,17 @@ public int getNumDocs() {
     return intValues;
   }
 
+  /**
+   * Get the int values for a single-valued column.
+   *
+   * @param column Column name
+   * @param evaluator JSON path evaluator
+   * @param buffer values to fill
+   */
+  public void fillValues(String column, JsonPathEvaluator evaluator, int[] buffer) {
```

Comment:
Does it make sense to have a more generic evaluator abstraction (with JsonPathEvaluator as one of the impls) here?

Reply:
I believe in the rule of 3: create a generic abstraction once you have three implementations.

Reply:
There is no external impact of changing `JsonPathEvaluator` to `Evaluator` which could be a regular expression (or something else) later, it all stays in pinot-core.

Reply:
Yeah I see your point about the rule of 3. I think I put the comment in a less effective place, since this is not an interface implementation.

Reply:
OK. I think what’s important about this API is the _shape_ of the result, and JsonPath evaluation will only ever produce a single vector of values with the same row indexing, in the same way that, say, a pushed down regular expression would. There are other obvious shapes to consider, such as an aggregator which produces a single value, or potentially something like a JSON UNNEST function which produces a matrix of values, typically with different row indexing. So any early abstraction for this needs to be able to model shape adequately. JsonPathEvaluator could be generalised to _TransformEvaluator_ because it preserves the row indexing.

Reply:
I just pushed something which should address your concerns.